### PR TITLE
feat: Prometheus export.go update for dynamic configs

### PR DESCRIPTION
**Pull request by TheSpiritXIII**
Pre-req: https://github.com/GoogleCloudPlatform/prometheus/pull/180

### CODE REVIEW

File: cmd/prometheus/main.go [hunk 1]
```diff
@@ -871,7 +881,7 @@ func main() {
 			name: "gcm_export",
 			reloader: func(cfg *config.Config) error {
 				// Call in closure to not call Global() before it's initialized below.
-				return gcm_export.Global().ApplyConfig(cfg)
+				return gcm_export.Global().ApplyConfig(cfg, nil)
```
Comment (pintohutch) [inline cmd/prometheus/main.go]: question: Don't you need to pass in the `exporterOpts` on reload instead of `nil`?

File: cmd/prometheus/main.go [hunk 2]
```diff
@@ -938,8 +948,17 @@ func main() {
 		)
 	}
 	{
+		exporterLogger := log.With(logger, "component", "gcm_exporter")
```
Comment (pintohutch) [inline cmd/prometheus/main.go:947]: question: Is there any way to wrap this up in a constructor, i.e. `newExporter` call?

It looks like the client (i.e. Prometheus) has a bit more explicit setup work to do now than before, which is less than ideal in a world where we're trying to defork ourselves. Is it feasible to consolidate these in a constructor in our export package?

A lot of these functions look new to me and I'm not exactly sure how they work, so I may just be out of the loop.

**Review by pintohutch — COMMENTED** [review]
Just giving my feedback - not blocking. I'll let @bwplotka weigh in.

File: cmd/prometheus/main.go (hunk 1 above)
Comment (TheSpiritXIII) [reply cmd/prometheus/main.go]: It's optional but I added it. Thanks!

File: cmd/prometheus/main.go (hunk 2 above)
Comment (TheSpiritXIII) [reply cmd/prometheus/main.go:947]: Done. Thanks!

File: go.mod [hunk 3]
```diff
@@ -1,14 +1,14 @@
 module github.com/prometheus/prometheus
 
-go 1.21
+go 1.22.3
```
Comment (pintohutch) [inline go.mod:3]: question (non-blocking): Is this change related?

I have nothing against bumping go for the module. But if it's not necessary, doing this in a separate PR or commit will make it easier to isolate these changes as they roll out.

File: cmd/prometheus/main.go [hunk 4]
```diff
@@ -871,7 +877,7 @@ func main() {
 			name: "gcm_export",
 			reloader: func(cfg *config.Config) error {
 				// Call in closure to not call Global() before it's initialized below.
-				return gcm_export.Global().ApplyConfig(cfg)
+				return gcm_export.Global().ApplyConfig(cfg, &opts.ExporterOpts)
```
Comment (pintohutch) [inline cmd/prometheus/main.go:880]: question (non-blocking): Ah - IIUC you originally left the arg `nil` as the reload function would never trigger at runtime anyway, since the opts (as of this PR) can only change when the process restarts. Do I have that correct?

**Review by pintohutch — APPROVED** [review]
LGTM % 2 non-blocking questions

File: cmd/prometheus/main.go (hunk 4 above)
Comment (TheSpiritXIII) [reply cmd/prometheus/main.go:880]: That's correct!

File: go.mod (hunk 3 above)
Comment (TheSpiritXIII) [reply go.mod:3]: `go mod tidy` updates the Go version to match the smaller of the minimum Go version of dependencies OR the current version in `go.mod`.

In this case, it updated to because prometheus-engine was using this version: https://github.com/GoogleCloudPlatform/prometheus-engine/blob/main/go.mod
